File: 백준/Silver/9012. 괄호/괄호.py
t = int(input())

for i in range(t):
    b=input()
    r=len(b)
    c=0
    if r % 2 == 1 or b[0] == ')' or b[-1] == '(':
        print("NO")
    else:
        for j in b:
            if j=='(':
                c+=1
            else:
                c-=1
                if c < 0:
                    break
        if c == 0:
            print("YES")
        else:
            print("NO")

File: 백준/Silver/9012. 괄호/괄호.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
   
        int T = Integer.parseInt(br.readLine());
        
        for(int i = 0; i < T; i++) {
            Stack<Character> stack = new Stack<>();
            String s = br.readLine();
            
            boolean check = true;
            
            for (int j = 0; j < s.length(); j++) {
                char c = s.charAt(j);
                if (c == '(') {
                    stack.push(c);
                } else if (stack.isEmpty()){
                    check = false;
                    break;
                } else {
                    stack.pop();
                }
            }
            System.out.println(check && stack.isEmpty() ? "YES" : "NO");
        }

        br.close();
    }
}